File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(int n) {
    vector<int> answer;
    
    //모든 칸의 개수 구하기
    int N = 0;
    for (int i=1 ; i<=n ; i++){
        N += i ;
    }
    
    vector<int> snail (N+1,-1); // 삼각형 초기화
    
    int cnt = 1; // 채워줘야 하는 수
    int idx = 1; // 현재 몇번째 칸인지
    int h = 1; // 현재 몇층인지
    bool upper = false; // 움직이는 방향
    
    while(cnt <= N){
        //현재 인덱스 숫자 채워주기
        snail[idx] = cnt;
        //왼쪽 아랫칸이 인덱스를 초과하지 않고 비어있다면
        if(!upper && idx + h <= N && snail[idx+h] == -1){
            idx += h;
            h++;
        }
        else{
            //바로 오른쪽 칸이 인덱스를 초과하지 않고 비어있다면
            if(idx + 1 <=N && snail[idx+1] == -1){
                idx ++;
            }
            else{
                //왼쪽 위 칸이 인덱스를 초과하지 않고 비어있다면
                if(0 < idx - h && snail[idx-h] == -1){
                    idx -= h;
                    h--;
                    upper = true;
                }else{
                    //올라가던 도중에 숫자가 있는칸을 만나면 다시 아래로 선회해주어야 한다.
                    if(snail[idx-h] != -1){
                        idx += h;
                        h++;
                        upper = false;
                    }
                }
            }
        }
        cnt ++;
    }
    
    for(int i = 1 ;i < snail.size() ;i++){
        answer.push_back(snail[i]);
    }
        
    return answer;
}

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    answer = []
    
    
    ## 삼각형 구조의 2차원 배열 만들기 
    
    snail = [[0]*i for i in range(1,n+1)]
    ##print(snail)
    
    num = 0
    r,c = -1,0
    for i in range (n):
        for j in range(i, n):
            num += 1
            
            if i % 3 == 0 :
                r += 1
            elif i % 3 == 1 : 
                c += 1
            elif i % 3 == 2 : 
                r -= 1 
                c -= 1 
            
            snail[r][c] = num
    
    # for row in snail:
    #     for value in row:
    #         answer.append(value)
            
            
    for i in range(0,n):
        for j in range(0,i+1):
            answer.append(snail[i][j])

    ##print(snail)
    
    
    
    return answer